helper/config/decode.go

```diff
@@ -179,10 +203,15 @@ func Decode(target interface{}, config *DecodeOpts, raws ...interface{}) error {
 
 				err = multierror.Append(err, unusedErr)
 			}
+			if err != nil {
+				return err
+			}
 		}
-		if err != nil {
-			return err
-		}
+	}
+
+	// Set the metadata if it is set
+	if config.Metadata != nil {
+		*config.Metadata = md
```

Comment:
Same here, decoder.Decode addresses things in `config` using pointers so this should not be necessary.

Reply:
removing the pointer causes `helper/config/decode.go:215:19: cannot use md (type mapstructure.Metadata) as type *mapstructure.Metadata in assignment`

Reply:
Most of this code was just moved a tad, not actually modified; we can refactor what was already there, like this assignment, but it feels a little out of scope.

Reply:
Oh right I see that it was simply moved, I just can't seem to understand why, is it because md is only fully know after the rest ?